### PR TITLE
✨ Logarithm, naturally

### PR DESCRIPTION
**Pull request by recmo**
## Description

Describe the changes made in your pull request here.

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [x] Ran `forge snapshot`?
- [x] Ran `npm run lint`?
- [x] Ran `forge test`?

_Pull requests with an incomplete checklist will be thrown out._

### CODE REVIEW

File: src/utils/BitwiseLib.sol [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+            r := shl(7, lt(0xffffffffffffffffffffffffffffffff, x))
+            r := or(r, shl(6, lt(0xffffffffffffffff, shr(r, x))))
+            r := or(r, shl(5, lt(0xffffffff, shr(r, x))))
+            r := or(r, shl(4, lt(0xffff, shr(r, x))))
+            r := or(r, shl(3, lt(0xff, shr(r, x))))
+            r := or(r, shl(2, lt(0xf, shr(r, x))))
+            r := or(r, shl(1, lt(0x3, shr(r, x))))
+            r := or(r, lt(0x1, shr(r, x)))
+        }
+    }
```
Comment (recmo) [inline src/utils/BitwiseLib.sol:23]: We can add more fun stuff here. Popcount, bitreverse, etc. See <https://graphics.stanford.edu/~seander/bithacks.html>

File: src/utils/FixedPointMathLib.sol [hunk 2]
```diff
@@ -27,6 +33,73 @@ library FixedPointMathLib {
...
+    // Computes ln(x) in 1e18 fixed point.
+    // Reverts if x is negative or zero.
+    // @param x The number to compute ln(x) for.
+    // @returns Approximately ln(x / 1e18) * 1e18 if x is strictly positive,
+    //   reverts with `LnNegativeUndefined()` if x is negative, and with
+    //   `Overflow()` if x is zero.
+    // Consumes about 670 gas.
+    function lnWad(int256 x) internal returns (int256 r) {
+        unchecked {
+            if (x < 1) {
```
Comment (transmissions11) [inline src/utils/FixedPointMathLib.sol:45]: would be slightly cheaper to use <= 0 cuz of the if -> JUMPI transform

File: src/utils/BitwiseLib.sol [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+/// @notice Bitwise library with operations for bit strings and integers.
+library BitwiseLib {
+    // Integer log2
+    // Closely related to count-leading-zeros.
+    // See <https://en.wikipedia.org/wiki/Find_first_set>
+    // @returns floor(log2(x)) if x is nonzero, otherwise 0.
+    //          This is the same as the location of the highest set bit.
+    // Consumes 232 gas. This could have been an 3 gas EVM opcode though.
+    function ilog2(uint256 x) internal returns (uint256 r) {
```
Comment (cruzdanilo) [inline src/utils/BitwiseLib.sol:12]: this function can be pure.

File: src/utils/FixedPointMathLib.sol [hunk 4]
```diff
@@ -27,6 +33,73 @@ library FixedPointMathLib {
...
     }
 
+    // Computes ln(x) in 1e18 fixed point.
+    // Reverts if x is negative or zero.
+    // @param x The number to compute ln(x) for.
+    // @returns Approximately ln(x / 1e18) * 1e18 if x is strictly positive,
+    //   reverts with `LnNegativeUndefined()` if x is negative, and with
+    //   `Overflow()` if x is zero.
+    // Consumes about 670 gas.
+    function lnWad(int256 x) internal returns (int256 r) {
```
Comment (cruzdanilo) [inline src/utils/FixedPointMathLib.sol:43]: this function can be pure.

**Comment by duncancmt** [comment]
The technique used in this PR is really different (and more efficient) compared to other libraries I've seen for calculating logarithms on the EVM. Is there a reference for the algorithm/technique used here and in #171 that I can read to understand better? Is it just TAOCP?

**Comment by transmissions11** [comment]
@duncancmt check out https://xn--2-umb.com/22/approximation/ and the notebooks in https://github.com/recmo/experiment-solexp @duncancmt

**Pull request merged**